src/models/logits.cpp

```diff
@@ -188,19 +236,41 @@ DeviceSpan<float> Logits::Get() {
 
     auto batched_logits_cpu = cpu_span<float>{cpu_tensor, element_count};
     HandleEOSArray(batched_logits_cpu);
+    if (!logits_masks_.empty()) {
+      AddMask(batched_logits_cpu, logits_masks_);
+    }
 
     logits_ = WrapTensor<float>(*state_.params_->p_device, *value32_cpu_);
     return logits_;
   }
 #endif
 
   HandleEOSArray(logits_.Span());
+  if (!logits_masks_.empty()) {
+    AddMask(logits_.Span(), logits_masks_);
+  }
   return logits_;
 }
 
 #pragma warning(pop)
 
-void Logits::Update() {
+void Logits::Update(DeviceSpan<int32_t> next_tokens_unk) {
+  if (!logits_processors_.empty() && logits_processors_.at(0)) {
+    auto next_tokens = next_tokens_unk.CopyDeviceToCpu();
+    for (int i = 0; i < next_tokens.size(); i++) {
+      logits_processors_[i]->CommitTokens(static_cast<uint32_t>(next_tokens[i]));
```

Comment:
`CommitTokens` can throw. Should the exception be handle here?

Reply:
I would like to let the exception block the whole execution for now. Because validated(done in the parser) grammar should work fine in one generation, leave it to throw exception should be good to optimize constrained decoding. And if this exception occurs often, we might consider to handle exception here and roll back to normal decoding in the future.